src/components/tl/cuda/tl_cuda.h

```diff
@@ -0,0 +1,189 @@
+        if (ucc_unlikely(cudaSuccess != e)) {                                  \
+            tl_error(_lib, "CUDA error %d %s", e, cudaGetErrorName(e));        \
+            _status = UCC_ERR_NO_MESSAGE;                                      \
+            goto _label;                                                       \
+        }                                                                      \
+    } while (0)
+
+#define CUDACHECK_NORET(_cmd)                                                  \
+    do {                                                                       \
+        cudaError_t e = _cmd;                                                  \
+        if (ucc_unlikely(cudaSuccess != e)) {                                  \
+            ucc_error("CUDA error %d %s", e, cudaGetErrorName(e));             \
+        }                                                                      \
+    } while (0)
+
+typedef struct ucc_tl_cuda_iface {
+    ucc_tl_iface_t super;
+} ucc_tl_cuda_iface_t;
+
+extern ucc_tl_cuda_iface_t ucc_tl_cuda;
+
+typedef struct ucc_tl_cuda_lib_config {
+    ucc_tl_lib_config_t super;
+    uint32_t            max_concurrent;
+} ucc_tl_cuda_lib_config_t;
+
+typedef struct ucc_tl_cuda_context_config {
+    ucc_tl_context_config_t super;
+} ucc_tl_cuda_context_config_t;
+
+typedef struct ucc_tl_cuda_lib {
+    ucc_tl_lib_t             super;
+    ucc_tl_cuda_lib_config_t cfg;
+} ucc_tl_cuda_lib_t;
+UCC_CLASS_DECLARE(ucc_tl_cuda_lib_t, const ucc_base_lib_params_t *,
+                  const ucc_base_config_t *);
+
+typedef struct ucc_tl_cuda_context {
+    ucc_tl_context_t              super;
+    ucc_tl_cuda_context_config_t  cfg;
+    int                           device;
+    ucc_mpool_t                   req_mp;
+    tl_cuda_ep_hash_t            *ipc_cache;
+} ucc_tl_cuda_context_t;
+UCC_CLASS_DECLARE(ucc_tl_cuda_context_t, const ucc_base_context_params_t *,
+                  const ucc_base_config_t *);
+
+typedef uint32_t ucc_tl_cuda_sync_state_t;
+
+typedef struct ucc_tl_cuda_shm_barrier {
+    ucc_rank_t   size;
+    ucc_rank_t   count;
+    int          sense;
+    ucc_status_t state[UCC_TL_CUDA_MAX_PEERS];
+    int          local_sense[UCC_TL_CUDA_MAX_PEERS];
+} ucc_tl_cuda_shm_barrier_t;
+
+typedef struct ucc_tl_cuda_rank_id {
+    int device;
+    int shm;
+} ucc_tl_cuda_rank_id_t;
+
+typedef struct ucc_tl_cuda_sync_data {
+    cudaEvent_t ipc_event_remote;
+} ucc_tl_cuda_sync_data_t;
+
+typedef struct ucc_tl_cuda_mem_info {
+    void               *ptr;
+    size_t              length;
+    size_t              offset;
+    cudaIpcMemHandle_t  handle;
+} ucc_tl_cuda_mem_info_t;
+
+typedef struct ucc_tl_cuda_sync {
+    uint32_t                 seq_num[2];
+    ucc_tl_cuda_mem_info_t   mem_info_src;
+    ucc_tl_cuda_mem_info_t   mem_info_dst;
+    cudaEvent_t              ipc_event_local;
+    cudaIpcEventHandle_t     ev_handle;
+    ucc_tl_cuda_sync_data_t  data[1];
```

Comment:
why array of size "1"?

Reply:
it's actually of size TEAM_SIZE, but it's defined as an array of 1 element for convenience